src/prof_recent.c

```diff
@@ -446,8 +455,35 @@ prof_recent_alloc_max_ctl_write(tsd_t *tsd, ssize_t max) {
 	return old_max;
 }
 
+ssize_t
+prof_recent_alloc_max_ctl_write(tsd_t *tsd, ssize_t max) {
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_dump_mtx);
+	ssize_t old_max = prof_recent_alloc_max_write_locked(tsd, max);
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_dump_mtx);
+	return old_max;
+}
+
+static prof_recent_t *
+prof_recent_alloc_swap_and_clear_locked(tsd_t *tsd, ssize_t *dump_count,
+    ssize_t *dump_max) {
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_dump_mtx);
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
```

Comment:
Similarly, don't lock internally here. Move the call site to inside of the lock, and invoke this one under the `node == NULL` initialization branch.